### PR TITLE
MAINT: signal: Make freqz handling of multidim. arrays match 0.19.1.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
I'll add some tests for broadcasting to the PR soon.

**Comment by josef-pkt** [comment]
A semi-random question, given that this is currently undefined in master: 

What is the multivariate behavior here? Or is it just vectorized univariate?

When I looked at this (9? years ago) there was no MIMO support in scipy.signal. I tried to figure out the freqz equivalent for the MIMO case. 
(AFAIR, I managed to get it to work for the num case but the denom case didn't have a nice equivalent of dividing by the fft.)

**Comment by WarrenWeckesser** [comment]
> What is the multivariate behavior here? Or is it just vectorized univariate?

It is vectorized univariate.

**Comment by larsoner** [comment]
> I'll add some tests for broadcasting to the PR soon.

I assume the plan is to make a bunch of cases that pass/fail on 0.19.1 and ensure they still pass/fail with this PR?

**Comment by WarrenWeckesser** [comment]
The tests with broadcasting that I am adding should pass with this PR and with 0.19.1.

**Comment by WarrenWeckesser** [comment]
I updated the pull request with unit tests and an edited docstring.

The test failure with numpy master is not related to this pull request; see https://github.com/numpy/numpy/issues/9734.

File: scipy/signal/filter_design.py [hunk 1]
```diff
@@ -362,37 +362,40 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
...
+                h = np.rollaxis(h, 0, h.ndim)
     del worN
 
-    if w.ndim != 1:
-        raise ValueError('w must be 1D')
     if h is None:  # still need to compute using freqs w
-        if w.size == 0:
-            raise ValueError('w must have at least one element, got 0')
+        #if w.size == 0:
+        #    raise ValueError('w must have at least one element, got 0')
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: delete?

Comment (larsoner) [reply scipy/signal/filter_design.py]: ... and actually `worN=[]` and `worN=0` should probably be tested too

File: scipy/signal/filter_design.py [hunk 2]
```diff
@@ -362,37 +362,40 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
     except TypeError:  # not int-like
         w = atleast_1d(worN)
     else:
-        if worN <= 0:
+        if worN < 0:
             raise ValueError('worN must be positive, got %s' % (worN,))
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: this should now say `non-negative`

File: scipy/signal/filter_design.py [hunk 3]
```diff
@@ -260,26 +260,29 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
...
+              A(e  )    a[0] + a[1]e    + ... + a[N]e
 
     Parameters
     ----------
     b : array_like
-        Numerator of a linear filter. Must be 1D.
+        Numerator of a linear filter.
     a : array_like
-        Denominator of a linear filter. Must be 1D.
+        Denominator of a linear filter.
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: The docstring will need to be updated to describe the broadcasting behavior of `b, a, worN`, i.e. if I have multidimensional versions of all of these how do I get them to broadcast properly?

Comment (larsoner) [reply scipy/signal/filter_design.py]: (and if it's complicated, adding a small example might be justified)

**Review by larsoner — COMMENTED** [review]
Otherwise LGTM, the changes are much simpler than I expected!

File: scipy/signal/filter_design.py (hunk 1 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: Yup and yup.

File: scipy/signal/filter_design.py (hunk 3 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: See if the updated version looks sufficient.

File: scipy/signal/filter_design.py (hunk 2 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: Yup, done.

File: scipy/signal/filter_design.py (hunk 1 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: I just pushed an update.  There are now a few tests where `worN` is 0 or an array with size 0.

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge

We should backport this to 1.0.0, right?

**Comment by rgommers** [comment]
> We should backport this to 1.0.0, right?

Yes. Just labeling it with ``backport-candidate`` does that job, I'll do all backports in one go.

**Comment by larsoner** [comment]
I'll merge early next week unless someone else wants to take a look

**Comment by pv** [comment]
The refguide checker found some issues in the examples:
https://travis-ci.org/scipy/scipy/jobs/278336980#L1761

**Comment by WarrenWeckesser** [comment]
The refguide checks are fixed in the last update.

**Comment by larsoner** [comment]
I'll merge tomorrow unless someone else wants to have a look

**Comment by larsoner** [comment]
Thanks @WarrenWeckesser

**Comment by WarrenWeckesser** [comment]
Thanks Eric, and sorry for leading you around in circles in the original pull request ("Transform the denominator!  No, don't transform the denominator!  `freqz` doesn't broadcast already!  Hold on, `freqz` *does* broadcast!").

**Comment by larsoner** [comment]
Not a problem, I'm glad we'll have it working, well documented, forward-compatible, *and* backward-compatible for the next release. Sometimes this requires some circling :)

**Comment by rgommers** [comment]
Backport in gh-7938

**Pull request merged**